beam-core/Database/Beam/Schema/Indices.hs

```diff
@@ -0,0 +1,357 @@
+
+instance (Beamable (PrimaryKey table),
+          IsNotEmptyData "Primary key" (PrimaryKey table Identity)) =>
+         FieldIndexBuilder (PrimaryKey table (TableField table')) where
+    buildFieldIndex =
+        fromList . allBeamValues (\(Columnar' (TableField fieldNm)) -> fieldNm)
+
+instance (Beamable (PrimaryKey table),
+          IsNotEmptyData "Primary key" (PrimaryKey table Identity)) =>
+         FieldIndexBuilder (PrimaryKey table (Nullable (TableField table'))) where
+    buildFieldIndex =
+        fromList . allBeamValues (\(Columnar' (TableField fieldNm)) -> fieldNm)
+
+-- | Gathers index fields from a user-supplied pack of table fields.
+class IndexBuilder table a where
+    buildIndex :: TableSettings table -> a -> NonEmpty Text
+
+-- | Field accessors are building blocks for indices.
+instance (f ~ TableField table, table ~ table', FieldIndexBuilder field) =>
+         IndexBuilder table' (table f -> field) where
+    buildIndex settings getter =
+        buildFieldIndex $ getter settings
+
+instance (IndexBuilder table a, IndexBuilder table b) =>
+         IndexBuilder table (a, b) where
+    buildIndex settings (a, b) = buildIndex settings a <> buildIndex settings b
+
+instance (IndexBuilder table a, IndexBuilder table b, IndexBuilder table c) =>
+         IndexBuilder table (a, b, c) where
+    buildIndex settings (a, b, c) =
+        buildIndex settings a <> buildIndex settings b <> buildIndex settings c
+
+-- | Make a table index builder covering the specified fields.
+--   Basic usage is to pass a table field accessor or a tuple of them to this function.
+--   Currently, no more than 3 elements in a tuple are supported, but feel free to nest
+--   tuples.
+--   Order of fields is preserved: tuples expand straightforwardly, while primary keys
+--   expand to a list of fields in the same order as they are mentioned in the
+--   corresponding constructor.
+tableIndex
+    :: IndexBuilder table a
+    => a
+    -> IndexOptions
+    -> EntityIndices be db (TableEntity table)
+tableIndex builder idxOpts =
+    EntityIndices . DL.singleton $
+        \(DatabaseEntity (DatabaseTable tblName tblSettings)) ->
+            Index tblName $ TableIndex (buildIndex tblSettings builder) idxOpts
+
+-- | For the given part of table @tblp@, indices derived from it.
+type TableIndicesBuilder tblp = DList (tblp -> TableIndex)
+
+-- | Helper for GAutoTableIndices
+contramapTableIndicesBuilder :: (b -> a) -> TableIndicesBuilder a -> TableIndicesBuilder b
+contramapTableIndicesBuilder f = fmap (. f)
+
+-- * Automatic indices definition
+
+-- | Indicates a reference from table @tbl@ to table @tbl'@.
+data (tbl :: (* -> *) -> *) :-> (tbl' :: (* -> *) -> *)
+
+-- | Provide options for an automatically created index, which is caused by a primary key
+-- of one table being embedded into another.
+-- This typeclass is only needed to be defined for inter-table references (see ':->').
+class IndexFromReference reference where
```

Comment:
The issue I see right now is that tables should be able to have two foreign keys that refer to the same table type at the haskell level, but really refer to two different tables at the database level.

For example

```haskell
data TwoKeyT f =
  TwoKey {
    _key1 :: PrimaryKey RefT f
    _key2 :: PrimaryKey RefT f
  }

data RefT f = Ref { _field1 :: C f Text, _field2 :: C f Int }

data Database entity = Database { _ref1 :: entity (TableEntity RefT), _ref2 :: entity (TableEntity RefT), _twoKey :: entity (TableEntity TwoKeyT) }
```

How would I make it so that it knows that _key1 refers to _ref2 and _key2 to _ref1 (or vice versa). I believe the way it is now, each type is uniquely linked to one database table, which is a large constraint.

One way to get around this that would also (I think) have the benefit of removing the instance overlap is to not use the type class mechanism and require the user to construct an appropriate value.

Reply:
I agree this is a major problem. In case of this implementation via type class mechanism a workaround is to parametrize table type with a phantom type parameter, thus making different tables of the same type distinguishable in a Haskell code.
However, making user explicitly make a construct which describes options for the generated indices seems to be virtually better in several ways. Doing so may be tricky if we still want the type system to enforce the user to specify indices for all embedded `PrimaryKey`s present; I described my raw suggestion right below this code, we could discuss it :)